tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala

```diff
@@ -97,52 +126,70 @@ final class ScalaJSClassEmitter(semantics: Semantics, outputMode: OutputMode,
+          Assign(Select(This()(tpe), name)(ftpe), zeroOf(ftpe)),
+          semantics, outputMode)
     }
 
-    {
-      implicit val pos = tree.pos
-      val typeVar = encodeClassVar(className)
-      val docComment = js.DocComment("@constructor")
-      val ctorDef = envFieldDef("c", className, ctorFun)
+    outputMode match {
+      case OutputMode.ECMAScript51Global | OutputMode.ECMAScript51Isolated =>
+        val ctorFun = {
+          val superCtorCall = tree.superClass.fold[js.Tree] {
+            js.Skip()(tree.pos)
+          } { parentIdent =>
+            implicit val pos = tree.pos
+            js.Apply(
+                js.DotSelect(encodeClassVar(parentIdent.name), js.Ident("call")),
+                List(js.This()))
+          }
+          js.Function(Nil,
+              js.Block(superCtorCall :: fieldDefs)(tree.pos))(tree.pos)
+        }
+
+        {
+          implicit val pos = tree.pos
+          val typeVar = encodeClassVar(className)
+          val docComment = js.DocComment("@constructor")
+          val ctorDef = envFieldDef("c", className, ctorFun)
 
-      val chainProto = tree.superClass.fold[js.Tree] {
-        js.Skip()
-      } { parentIdent =>
-        js.Block(
-          js.Assign(typeVar.prototype,
-              js.New(envField("h", parentIdent.name, parentIdent.originalName), Nil)),
-          genAddToPrototype(className, js.Ident("constructor"), typeVar)
-        )
-      }
+          val chainProto = tree.superClass.fold[js.Tree] {
+            js.Skip()
+          } { parentIdent =>
+            js.Block(
+              js.Assign(typeVar.prototype,
+                  js.New(envField("h", parentIdent.name, parentIdent.originalName), Nil)),
+              genAddToPrototype(className, js.Ident("constructor"), typeVar)
+            )
+          }
 
-      val inheritableCtorDef = {
-        js.Block(
-          js.DocComment("@constructor"),
-          envFieldDef("h", className, js.Function(Nil, js.Skip())),
-          js.Assign(envField("h", className).prototype, typeVar.prototype)
-        )
-      }
+          val inheritableCtorDef = {
+            js.Block(
+              js.DocComment("@constructor"),
+              envFieldDef("h", className, js.Function(Nil, js.Skip())),
+              js.Assign(envField("h", className).prototype, typeVar.prototype)
+            )
+          }
 
-      js.Block(docComment, ctorDef, chainProto, inheritableCtorDef)
+          js.Block(docComment, ctorDef, chainProto, inheritableCtorDef)
+        }
+
+      case OutputMode.ECMAScript6 =>
+        implicit val pos = tree.pos
+        if (fieldDefs.isEmpty) {
+          js.Skip()
+        } else {
+          val superCtorCall = tree.superClass.fold[js.Tree] {
+            js.Skip()(tree.pos)
+          } { parentIdent =>
+            js.Apply(js.Super(), Nil)
+          }
+          js.MethodDef(static = false, js.Ident("constructor"), Nil,
+              js.Block(superCtorCall :: fieldDefs))
+        }
```

Comment:
Consider refactoring this in `genES5Constructor`, `genES6Constructor` and `genFieldDefs`. I think the method is really too long like this.